File: packages/coordinator/src/profiles/control.ts
/**
 * control/1.0 profile (profiles/control.md).
 *
 * Methods:
 *   - control.pause / control.resume   (scope: task|participant|workspace)
 *   - control.cancel
 *   - control.snapshot                  (returns art_... id)
 *   - control.rollback                  (to_snapshot_artefact_id + what_to_restore)
 *   - control.supersede                 (creates successor in one shot)
 *   - control.set_mode_ceiling
 */
import type { Coordinator } from "../coordinator.js";
import { E, rpcError } from "../jsonrpc.js";
import type { Mode, SnapshotArtefact, Task } from "../types.js";

const VALID_MODES = new Set<Mode>(["shadow", "trial", "production"]);
const VALID_SCOPES = new Set(["task", "participant", "workspace"]);

export function registerControl(coord: Coordinator): void {
  coord.handlers.set("control.pause", (p) => {
    const ws = coord.workspaces.get(p.workspace as string);
    if (!ws) return { error: rpcError(E.PARAMS, "Unknown workspace") };
    const scope = (p.scope as string) || "task";
    if (!VALID_SCOPES.has(scope)) {
      return { error: rpcError(E.PARAMS, `scope must be task/participant/workspace`) };
    }
    if (scope === "task") {
      const task = ws.tasks.get(p.task_id as string);
      if (!task) return { error: rpcError(E.PARAMS, "Unknown task") };
      if (task.state === "completed" || task.state === "declined" || task.state === "cancelled") {
        return { error: rpcError(E.CONTROL_NOT_AUTHORISED, `Cannot pause ${task.state} task`) };
      }
      task.paused = true;
      const prior = task.state;
      task.state = "paused";
      task.updated_at = coord.now();
      task.history.push({ ts: task.updated_at, from: p.from as string, state: "paused", note: `was ${prior}` });
      return { result: { scope: "task", task_id: task.id, state: "paused" } };
    }
    if (scope === "participant") {
      const uri = (p.participant_uri as string) || (p.uri as string);
      if (!uri || !ws.members.has(uri)) return { error: rpcError(E.PARAMS, `Unknown participant: ${uri}`) };
      ws.members.get(uri)!.paused = true;
      return { result: {
        scope: "participant", participant_uri: uri, paused: true,
        in_flight_policy: (p.in_flight_policy as string) || "allow_to_complete",
      }};
    }
    // workspace
    ws.state = "paused";
    return { result: { scope: "workspace", workspace: ws.id, state: "paused" } };
  });

  coord.handlers.set("control.resume", (p) => {
    const ws = coord.workspaces.get(p.workspace as string);
    if (!ws) return { error: rpcError(E.PARAMS, "Unknown workspace") };
    const scope = (p.scope as string) || "task";
    if (!VALID_SCOPES.has(scope)) {
      return { error: rpcError(E.PARAMS, `scope must be task/participant/workspace`) };
    }
    if (scope === "task") {
      const task = ws.tasks.get(p.task_id as string);
      if (!task) return { error: rpcError(E.PARAMS, "Unknown task") };
      if (task.state !== "paused") {
        return { error: rpcError(E.CONTROL_NOT_AUTHORISED, `Cannot resume ${task.state} task`) };
      }
      task.paused = false;
      task.state = "in_progress";
      task.updated_at = coord.now();
      task.history.push({ ts: task.updated_at, from: p.from as string, state: "in_progress" });
      return { result: { scope: "task", task_id: task.id, state: "in_progress" } };
    }
    if (scope === "participant") {
      const uri = (p.participant_uri as string) || (p.uri as string);
      if (!uri || !ws.members.has(uri)) return { error: rpcError(E.PARAMS, `Unknown participant: ${uri}`) };
      ws.members.get(uri)!.paused = false;
      return { result: { scope: "participant", participant_uri: uri, paused: false } };
    }
    ws.state = "active";
    return { result: { scope: "workspace", workspace: ws.id, state: "active" } };
  });

  coord.handlers.set("control.cancel", (p) => {
    const ws = coord.workspaces.get(p.workspace as string);
    if (!ws) return { error: rpcError(E.PARAMS, "Unknown workspace") };
    const task = ws.tasks.get(p.task_id as string);
    if (!task) return { error: rpcError(E.PARAMS, "Unknown task") };
    if (task.state === "completed" || task.state === "declined" || task.state === "cancelled") {
      return { error: rpcError(E.CONTROL_NOT_AUTHORISED, `Cannot cancel ${task.state} task`) };
    }
    task.state = "cancelled";
    task.updated_at = coord.now();
    task.history.push({ ts: task.updated_at, from: p.from as string, state: "cancelled", note: p.reason as string | undefined });
    return { result: { state: "cancelled" } };
  });

  coord.handlers.set("control.snapshot", (p) => {
    const ws = coord.workspaces.get(p.workspace as string);
    if (!ws) return { error: rpcError(E.PARAMS, "Unknown workspace") };
    const include: string[] = (p.include as string[]) ?? ["members", "open_tasks", "mode_ceiling"];
    const state: Record<string, unknown> = {};
    if (include.includes("members")) {
      state.members = Array.from(ws.members.values()).map(m => ({
        uri: m.uri, type: m.type, role: m.role, scopes: m.scopes,
      }));
    }
    if (include.includes("open_tasks")) {
      state.open_tasks = Array.from(ws.tasks.values())
        .filter(t => t.state !== "completed" && t.state !== "declined" && t.state !== "cancelled")
        .map(t => ({ id: t.id, kind: t.kind, state: t.state, assignee: t.assignee }));
    }
    if (include.includes("mode_ceiling")) state.mode_ceiling = ws.mode_ceiling;
    if (include.includes("policy")) state.routing_policy_uri = ws.routing_policy_uri;
    if (include.includes("audit") || include.includes("all")) state.audit_seq = ws.audit.length;

    const snapId = coord.ids.artefactId();
    const snap: SnapshotArtefact = {
      id: snapId,
      kind: "snapshot",
      ts: coord.now(),
      by: p.from as string,
      workspace: ws.id,
      audit_seq: ws.audit.length,
      label: p.label as string | undefined,
      include,
      state,
    };
    ws.snapshots.set(snapId, snap);
    return { result: { snapshot_artefact_id: snapId, audit_seq: snap.audit_seq, artefact: snap } };
  });

  coord.handlers.set("control.rollback", (p) => {
    const ws = coord.workspaces.get(p.workspace as string);
    if (!ws) return { error: rpcError(E.PARAMS, "Unknown workspace") };
    const snapId = (p.to_snapshot_artefact_id as string)
                 || (p.snapshot_artefact_id as string)
                 || (p.snapshot_id as string);
    if (!snapId) return { error: rpcError(E.PARAMS, "to_snapshot_artefact_id is required") };
    const snap = ws.snapshots.get(snapId);
    if (!snap) return { error: rpcError(E.CONTROL_SNAPSHOT_NOT_FOUND, `Unknown snapshot: ${snapId}`) };
    const what: string[] = (p.what_to_restore as string[]) ?? snap.include;
    const restored: string[] = [];
    if (what.includes("mode_ceiling") && snap.state.mode_ceiling) {
      ws.mode_ceiling = snap.state.mode_ceiling as Mode;
      restored.push("mode_ceiling");
    }
    if (what.includes("members") && Array.isArray(snap.state.members)) {
      const snapByUri = new Map<string, Record<string, unknown>>();
      for (const m of snap.state.members as Array<Record<string, unknown>>) {
        snapByUri.set(m.uri as string, m);
      }
      for (const [uri, snapM] of snapByUri) {
        if (ws.members.has(uri)) {
          const cur = ws.members.get(uri)!;
          cur.role = (snapM.role as string) ?? cur.role;
          cur.scopes = snapM.scopes as string[] | undefined;
        }
      }
      restored.push("members");
    }
    return { result: {
      rolled_back_to: snapId, audit_seq: snap.audit_seq, restored,
      reason: p.reason as string | undefined,
    }};
  });

  coord.handlers.set("control.supersede", (p) => {
    const ws = coord.workspaces.get(p.workspace as string);
    if (!ws) return { error: rpcError(E.PARAMS, "Unknown workspace") };
    const oldId = p.task_id as string;
    const old = ws.tasks.get(oldId);
    if (!old) return { error: rpcError(E.PARAMS, "Unknown task") };
    const newSpec = (p.successor_task as Record<string, unknown>) ?? {};
    if (!newSpec.kind) {
      return { error: rpcError(E.PARAMS, "successor_task must include kind") };
    }
    const assignee = (newSpec.assignee as string) || old.assignee;
    if (!ws.members.has(assignee)) {
      return { error: rpcError(E.PARAMS, "successor assignee not in workspace") };
    }
    const now = coord.now();
    const newId = coord.ids.taskId();
    const newTask: Task = {
      id: newId,
      kind: newSpec.kind as string,
      state: "created",
      assignee,
      delegator: p.from as string,
      input: (newSpec.input as Record<string, unknown>) ?? {},
      created_at: now,
      updated_at: now,
      mode: (newSpec.mode as Mode) ?? old.mode,
      supersedes: old.id,
      history: [{ ts: now, from: p.from as string, state: "created",
                  note: `supersedes ${old.id}: ${(p.reason as string) || ""}` }],
      paused: false,
    };
    ws.tasks.set(newId, newTask);
    old.state = "superseded";
    old.superseded_by = newId;
    old.updated_at = now;
    old.history.push({ ts: now, from: p.from as string, state: "superseded",
                       note: `-> ${newId}: ${(p.reason as string) || ""}` });
    return { result: { superseded_task_id: old.id, new_task_id: newId, reason: p.reason as string | undefined } };
  });

  coord.handlers.set("control.set_mode_ceiling", (p) => {
    const ws = coord.workspaces.get(p.workspace as string);
    if (!ws) return { error: rpcError(E.PARAMS, "Unknown workspace") };
    const m = (p.new_ceiling as Mode) ?? (p.mode_ceiling as Mode);
    if (!VALID_MODES.has(m)) {
      return { error: rpcError(E.PARAMS, `new_ceiling must be shadow/trial/production`) };
    }
    ws.mode_ceiling = m;
    return { result: { mode_ceiling: ws.mode_ceiling, reason: p.reason as string | undefined } };
  });
}

File: packages/coordinator-py/chap_coordinator/profiles/control.py
"""
chap_coordinator.profiles.control

The control/1.0 profile (profiles/control.md).

Methods (all per spec):
  - control.pause            (scope: task|participant|workspace)
  - control.resume
  - control.cancel
  - control.snapshot         (returns art_... id)
  - control.rollback         (takes to_snapshot_artefact_id + what_to_restore)
  - control.supersede        (creates successor_task in one shot)
  - control.set_mode_ceiling

All control.* methods are privileged. Step-up enforcement is handled
at the coordinator dispatch layer; we surface CONTROL_STEP_UP_REQUIRED
where appropriate.

Error codes:
  -32060 step-up required
  -32061 not authorised for control operations
  -32062 snapshot artefact not found
  -32063 workspace paused; operation blocked
"""
from __future__ import annotations

from typing import TYPE_CHECKING

from ..jsonrpc import E, rpc_error
from ..types import SnapshotArtefact, Task, TaskHistoryEntry

if TYPE_CHECKING:
    from ..coordinator import Coordinator


_VALID_MODES = ("shadow", "trial", "production")
_VALID_SCOPES = ("task", "participant", "workspace")


def register_control(coord: "Coordinator") -> None:

    def control_pause(p: dict) -> dict:
        ws = coord.workspaces.get(p.get("workspace", ""))
        if not ws:
            return {"error": rpc_error(E.PARAMS, "Unknown workspace")}
        scope = p.get("scope", "task")
        if scope not in _VALID_SCOPES:
            return {"error": rpc_error(E.PARAMS, f"scope must be one of {_VALID_SCOPES}")}

        if scope == "task":
            task = ws.tasks.get(p.get("task_id", ""))
            if not task:
                return {"error": rpc_error(E.PARAMS, "Unknown task")}
            if task.state in ("completed", "declined", "cancelled"):
                return {"error": rpc_error(E.CONTROL_NOT_AUTHORISED,
                                           f"Cannot pause {task.state} task")}
            task.paused = True
            prior = task.state
            task.state = "paused"
            task.updated_at = coord.now_iso()
            task.history.append(TaskHistoryEntry(
                ts=task.updated_at, from_=p.get("from", ""),
                state="paused", note=f"was {prior}",
            ))
            return {"result": {"scope": "task", "task_id": task.id,
                               "state": "paused"}}

        if scope == "participant":
            uri = p.get("participant_uri") or p.get("uri")
            if not uri or uri not in ws.members:
                return {"error": rpc_error(E.PARAMS,
                                           f"Unknown participant: {uri}")}
            ws.members[uri].paused = True
            return {"result": {"scope": "participant", "participant_uri": uri,
                               "paused": True,
                               "in_flight_policy": p.get("in_flight_policy", "allow_to_complete")}}

        # workspace
        ws.state = "paused"
        return {"result": {"scope": "workspace", "workspace": ws.id,
                           "state": "paused"}}

    def control_resume(p: dict) -> dict:
        ws = coord.workspaces.get(p.get("workspace", ""))
        if not ws:
            return {"error": rpc_error(E.PARAMS, "Unknown workspace")}
        scope = p.get("scope", "task")
        if scope not in _VALID_SCOPES:
            return {"error": rpc_error(E.PARAMS, f"scope must be one of {_VALID_SCOPES}")}

        if scope == "task":
            task = ws.tasks.get(p.get("task_id", ""))
            if not task:
                return {"error": rpc_error(E.PARAMS, "Unknown task")}
            if task.state != "paused":
                return {"error": rpc_error(E.CONTROL_NOT_AUTHORISED,
                                           f"Cannot resume {task.state} task")}
            task.paused = False
            task.state = "in_progress"
            task.updated_at = coord.now_iso()
            task.history.append(TaskHistoryEntry(
                ts=task.updated_at, from_=p.get("from", ""), state="in_progress",
            ))
            return {"result": {"scope": "task", "task_id": task.id,
                               "state": "in_progress"}}

        if scope == "participant":
            uri = p.get("participant_uri") or p.get("uri")
            if not uri or uri not in ws.members:
                return {"error": rpc_error(E.PARAMS,
                                           f"Unknown participant: {uri}")}
            ws.members[uri].paused = False
            return {"result": {"scope": "participant",
                               "participant_uri": uri, "paused": False}}

        # workspace
        ws.state = "active"
        return {"result": {"scope": "workspace", "workspace": ws.id,
                           "state": "active"}}

    def control_cancel(p: dict) -> dict:
        ws = coord.workspaces.get(p.get("workspace", ""))
        if not ws:
            return {"error": rpc_error(E.PARAMS, "Unknown workspace")}
        task = ws.tasks.get(p.get("task_id", ""))
        if not task:
            return {"error": rpc_error(E.PARAMS, "Unknown task")}
        if task.state in ("completed", "declined", "cancelled"):
            return {"error": rpc_error(E.CONTROL_NOT_AUTHORISED,
                                       f"Cannot cancel {task.state} task")}
        task.state = "cancelled"
        task.updated_at = coord.now_iso()
        task.history.append(TaskHistoryEntry(
            ts=task.updated_at, from_=p.get("from", ""),
            state="cancelled", note=p.get("reason"),
        ))
        return {"result": {"state": "cancelled"}}

    def control_snapshot(p: dict) -> dict:
        ws = coord.workspaces.get(p.get("workspace", ""))
        if not ws:
            return {"error": rpc_error(E.PARAMS, "Unknown workspace")}
        include = list(p.get("include") or ["members", "open_tasks",
                                            "mode_ceiling"])
        # Capture the requested slice of state
        state: dict = {}
        if "members" in include:
            state["members"] = [m.to_dict() for m in ws.members.values()]
        if "open_tasks" in include:
            state["open_tasks"] = [
                t.to_dict() for t in ws.tasks.values()
                if t.state not in ("completed", "declined", "cancelled")
            ]
        if "mode_ceiling" in include:
            state["mode_ceiling"] = ws.mode_ceiling
        if "policy" in include:
            state["routing_policy_uri"] = ws.routing_policy_uri
        if "all" in include or "audit" in include:
            state["audit_seq"] = len(ws.audit)

        snap_id = coord.ids.artefact_id()  # snapshots are artefacts per spec
        snap = SnapshotArtefact(
            id=snap_id,
            ts=coord.now_iso(),
            by=p.get("from", ""),
            workspace=ws.id,
            audit_seq=len(ws.audit),
            label=p.get("label"),
            include=include,
            state=state,
        )
        ws.snapshots[snap_id] = snap
        return {"result": {"snapshot_artefact_id": snap_id,
                           "audit_seq": snap.audit_seq,
                           "artefact": snap.to_dict()}}

    def control_rollback(p: dict) -> dict:
        ws = coord.workspaces.get(p.get("workspace", ""))
        if not ws:
            return {"error": rpc_error(E.PARAMS, "Unknown workspace")}
        snap_id = (p.get("to_snapshot_artefact_id")
                   or p.get("snapshot_artefact_id")
                   or p.get("snapshot_id"))
        if not snap_id:
            return {"error": rpc_error(E.PARAMS,
                                       "to_snapshot_artefact_id is required")}
        snap = ws.snapshots.get(snap_id)
        if not snap:
            return {"error": rpc_error(E.CONTROL_SNAPSHOT_NOT_FOUND,
                                       f"Unknown snapshot: {snap_id}")}
        what = list(p.get("what_to_restore") or snap.include)

        # Apply the rollback. Per spec, this APPENDS, it does not truncate.
        # We restore the named slices of state.
        restored: list[str] = []
        if "mode_ceiling" in what and "mode_ceiling" in snap.state:
            ws.mode_ceiling = snap.state["mode_ceiling"]
            restored.append("mode_ceiling")
        if "members" in what and "members" in snap.state:
            # Restore each member's role/scopes from the snapshot;
            # we don't recreate departed members (audit is append-only)
            snap_uris = {m["uri"]: m for m in snap.state["members"]}
            for uri, snap_m in snap_uris.items():
                if uri in ws.members:
                    ws.members[uri].role = snap_m.get("role",
                                                     ws.members[uri].role)
                    ws.members[uri].scopes = snap_m.get("scopes")
            restored.append("members")

        return {"result": {
            "rolled_back_to": snap_id,
            "audit_seq": snap.audit_seq,
            "restored": restored,
            "reason": p.get("reason"),
        }}

    def control_supersede(p: dict) -> dict:
        ws = coord.workspaces.get(p.get("workspace", ""))
        if not ws:
            return {"error": rpc_error(E.PARAMS, "Unknown workspace")}
        old_id = p.get("task_id")
        old = ws.tasks.get(old_id or "")
        if not old:
            return {"error": rpc_error(E.PARAMS, "Unknown task")}

        # successor_task is an object describing the new task; the Coordinator
        # creates it as part of the supersede operation (per spec).
        new_spec = p.get("successor_task") or {}
        if not new_spec or "kind" not in new_spec:
            return {"error": rpc_error(E.PARAMS,
                                       "successor_task must include kind")}
        assignee = new_spec.get("assignee") or old.assignee
        if assignee not in ws.members:
            return {"error": rpc_error(E.PARAMS,
                                       "successor assignee not in workspace")}

        now = coord.now_iso()
        new_id = coord.ids.task_id()
        new_task = Task(
            id=new_id,
            kind=new_spec["kind"],
            state="created",
            assignee=assignee,
            delegator=p.get("from", ""),
            input=new_spec.get("input") or {},
            created_at=now,
            updated_at=now,
            mode=new_spec.get("mode") or old.mode,
            supersedes=old.id,
            history=[TaskHistoryEntry(
                ts=now, from_=p.get("from", ""), state="created",
                note=f"supersedes {old.id}: {p.get('reason') or ''}",
            )],
        )
        ws.tasks[new_id] = new_task

        old.state = "superseded"
        old.superseded_by = new_id
        old.updated_at = now
        old.history.append(TaskHistoryEntry(
            ts=now, from_=p.get("from", ""), state="superseded",
            note=f"-> {new_id}: {p.get('reason') or ''}",
        ))
        return {"result": {
            "superseded_task_id": old.id,
            "new_task_id": new_id,
            "reason": p.get("reason"),
        }}

    def control_set_mode_ceiling(p: dict) -> dict:
        ws = coord.workspaces.get(p.get("workspace", ""))
        if not ws:
            return {"error": rpc_error(E.PARAMS, "Unknown workspace")}
        new_ceiling = p.get("new_ceiling") or p.get("mode_ceiling")
        if new_ceiling not in _VALID_MODES:
            return {"error": rpc_error(E.PARAMS,
                                       f"new_ceiling must be one of {_VALID_MODES}")}
        ws.mode_ceiling = new_ceiling
        return {"result": {"mode_ceiling": ws.mode_ceiling,
                           "reason": p.get("reason")}}

    coord._handlers["control.pause"] = control_pause
    coord._handlers["control.resume"] = control_resume
    coord._handlers["control.cancel"] = control_cancel
    coord._handlers["control.snapshot"] = control_snapshot
    coord._handlers["control.rollback"] = control_rollback
    coord._handlers["control.supersede"] = control_supersede
    coord._handlers["control.set_mode_ceiling"] = control_set_mode_ceiling
